core/src/es/event.rs

```diff
@@ -26,60 +33,75 @@ impl Revision {
         Some(Self(NonZeroU16::new(u16::try_from(value).ok()?)?))
     }
 
-    /// Creates a new [`Revision`] out of the given `value` without checking its
-    /// invariants.
+    /// Creates a new [`Version`] out of the provided `value` without checking
+    /// its invariants.
     ///
     /// # Safety
     ///
-    /// The given `value` must not be `0` (zero).
+    /// The provided `value` must not be `0` (zero).
     #[inline]
     #[must_use]
     pub const unsafe fn new_unchecked(value: u16) -> Self {
         // SAFETY: Safety invariants are the same as for this method.
         Self(unsafe { NonZeroU16::new_unchecked(value) })
     }
 
-    /// Returns the value of this [`Revision`] as a primitive type.
+    /// Returns the value of this [`Version`] as a primitive type.
     #[inline]
     #[must_use]
     pub const fn get(self) -> u16 {
         self.0.get()
     }
 }
 
-/// [`Event`] of a concrete [`Revision`].
-pub trait Revised {
-    /// [`Name`] of this [`Event`].
-    const NAME: Name;
-
-    /// [`Revision`] of this [`Event`].
-    const REVISION: Revision;
-}
-
 /// [Event Sourcing] event describing something that has occurred (happened
 /// fact).
 ///
 /// [Event Sourcing]: https://martinfowler.com/eaaDev/EventSourcing.html
 pub trait Event {
     /// Returns [`Name`] of this [`Event`].
     ///
-    /// _Note:_ This should effectively be a constant value, and should never
-    /// change.
+    /// > **NOTE:** This should effectively be a constant value, and should
+    /// >           never change.
     #[must_use]
     fn name(&self) -> Name;
+}
 
-    /// Returns [`Revision`] of this [`Event`].
-    #[must_use]
-    fn revision(&self) -> Revision;
+/// Concrete [`Event`] defined statically.
+pub trait Static {
+    /// Concrete [`Name`] of this [`Event`].
+    const NAME: Name;
 }
 
-impl<Ev: Revised + ?Sized> Event for Ev {
+impl<Ev: Static + ?Sized> Event for Ev {
     fn name(&self) -> Name {
-        <Self as Revised>::NAME
+        <Self as Static>::NAME
     }
+}
 
-    fn revision(&self) -> Revision {
-        <Self as Revised>::REVISION
+/// [`Event`] capable of evolving with time.
+pub trait Revisable: Event {
+    /// Type of this [`Event`]'s [`Revision`] number.
+    type Revision: Revision;
+
+    /// Returns [`Revision`] of this [`Event`].
+    #[must_use]
+    fn revision(&self) -> Self::Revision;
+}
+
+/// [`StaticEvent`] of a concrete [`Revision`].
+///
+/// [`StaticEvent`]: Static
+pub trait Concrete: Static + Revisable {
+    /// Concrete [`Revision`] of this [`Event`].
+    const REVISION: Self::Revision;
+}
+
+impl<Ev: Concrete + ?Sized> Revisable for Ev {
+    type Revision = Self::Revision;
```

Comment:
@tyranron This impl will not work due to evaluation overflow. `Concrete` and `Revisable` should be implemented separately.

Reply:
Also we should provide `revision_type` argument (or something similar) to `Event` macro, because Rust can't determine exact type we place into `REVISION`.

Reply:
Alternatively, we can do `trait Revisable<Revision>` to generalize over `Revision`, but this requires us to pass generic to `Revisable` each time.

Reply:
Solution for now:
- in derive macro use `std::num::NonZeroU16`

Solution for future:
- add `const fn event::Revision::new` when const traits are supported (performs validation and builds a value)
- add custom revision types for derive macro:
    - provide `#[event(revision_type = NonZeroU16)]` attribute argument
    - extend `revision` argument to `#[event(revision = NonZeroU16::new(1))]`